OrbitQt/servicedeploymanager.h

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_QT_SERVICE_DEPLOY_MANAGER_H_
+#define ORBIT_QT_SERVICE_DEPLOY_MANAGER_H_
+
+#include <QEventLoop>
+#include <QObject>
+#include <QPointer>
+#include <string>
+
+#include "OrbitSshQt/Session.h"
+#include "OrbitSshQt/SftpChannel.h"
+#include "OrbitSshQt/SftpOperation.h"
+#include "OrbitSshQt/Task.h"
+#include "OrbitSshQt/Tunnel.h"
+#include "deploymentconfigurations.h"
+
+namespace OrbitQt {
+
+class ServiceDeployManager : public QObject {
+  Q_OBJECT
+
+ public:
+  struct Ports {
+    uint16_t asio_port;
+    uint16_t grpc_port;
+  };
+
+  explicit ServiceDeployManager(
+      DeploymentConfiguration deployment_configuration,
+      OrbitSsh::Credentials creds, Ports remote_ports,
+      QObject* parent = nullptr);
+
+  outcome::result<Ports> exec();
+  void shutdown();
+
+  void cancel();
+
+ signals:
+  void statusMessage(const QString&);
+  void socketErrorOccured(std::error_code);
+
+ private:
+  QEventLoop loop;
+
+  DeploymentConfiguration deployment_configuration_;
+  OrbitSsh::Credentials credentials_;
+  Ports remote_ports_;
+  std::optional<OrbitSsh::Context> context_;
+  std::optional<OrbitSshQt::Session> session_;
+  std::optional<OrbitSshQt::Task> orbit_service_task_;
+  std::optional<OrbitSshQt::Tunnel> asio_tunnel_;
+  std::optional<OrbitSshQt::Tunnel> grpc_tunnel_;
+
+  outcome::result<void> connectToServer();
+  outcome::result<bool> checkIfInstalled();
+  outcome::result<void> copyOrbitServicePackage();
+  outcome::result<void> copyOrbitServiceExecutable();
+  outcome::result<void> installOrbitServicePackage();
+  outcome::result<void> startOrbitService();
+  outcome::result<void> startOrbitServicePrivileged();
+  outcome::result<uint16_t> startAsioTunnel();
+  outcome::result<uint16_t> startGrpcTunnel();
+
+  outcome::result<void> startSftpChannel(OrbitSshQt::SftpChannel&);
```

Comment:
maybe consider making those arguments pointers if const ref is not possible

Reply:
Done.

Reply:
Done.